### PR TITLE
Refactor Mosa.DeviceSystem (part 1)

### PR DESCRIPTION
**Pull request by AnErrupTion**
This series of PRs aims to refactor parts of the code base with the primary goal of making them easier to understand. Here are the most notable changes in part 1 of refactoring Mosa.DeviceSystem (and parts of Mosa.DeviceDriver):

- Throughout all refactored files, pointless XML comments have been removed, and per-class comments have been added, giving more useful insight into what the class is, and does. This allow for better general understanding of the code.
- Added useful properties in the EditorConfig file, and removed unused ones
- The StandardKeyboard driver (and the IKeyboardDevice interface) are now able to be requested scan codes via blocking and non-blocking ways. (Previously, it was only non-blocking, which could cause some issues in e.g. console environments)
- The ``_lock`` field in BaseDeviceDriver has been renamed to ``DriverLock``
- GraphicalTextDevice's (and thus, ITextDevice's) ClearScreen() method was renamed to Clear()
- The following, unused classes and interfaces have been removed **(NOTE: Many files appear as deleted in Git, when in fact they were simply moved. Please refer to this list which is exhaustive.)**:
  - IAudioDevice (unused, we have no audio subsystem yet)
  - IDMAChannel (unused + uninherited)
  - IDateTime (unused + uninherited)
  - IMultibootSpecification (unused + uninherited)
  - INetworkDevice (unused, we have no network subsystem yet)
  - IOPortResources (unused)
  - IPixelPaletteGraphicsDevice (unused)
  - ISerialDevice (unused + uninherited)
  - ITextScreen (it was only inherited by TextScreen, and there was no need for it even then)
  - ITimer (unused + uninherited)
  - MACAddress (same as INetworkDevice)
  - MemoryResources (unused)
  - IPCIDevice (it was only inherited by PCIDevice)
  - IPCIDeviceResource (same as IPCIDevice + it provided properties and methods that IPCIDevice already had)
  - SpinLock (unused)
- Mosa.DeviceSystem has been partly restructured (not finished yet, which is mostly the sole reason why this PR has been split into multiple parts). They contain the following classes and interfaces thus far:
  - Graphics
    - FrameBuffer32
    - IGraphicsDevice
  - Keyboard
    - IKeyboard
    - Key
    - KeyEvent
    - KeyType
    - Keyboard
    - IKeyboardDevice
    - IScanCodeMap
  - TextDevice
    - GraphicalTextDevicec
    - ITextDevice
    - TextScreen
  - PCI
    - IPCIController
    - IPCIControllerLegacy
  - VirtIO (nothing has changed there)
  - Fonts
    - ASC16Font
    - ISimpleFont
    - SimpleBitFont
  - Services (nothing has changed there either, apart from the name which previously was Service)

### CODE REVIEW

File: Source/.editorconfig [hunk 1]
```diff
@@ -11,24 +11,65 @@ indent_style = tab
...
 dotnet_style_prefer_auto_properties = true
 dotnet_style_prefer_compound_assignment = true
 dotnet_style_coalesce_expression = true
 dotnet_style_prefer_simplified_interpolation = true
-dotnet_diagnostic.CS0168.severity = none
 csharp_style_var_for_built_in_types = true
 csharp_style_var_when_type_is_apparent = true
 
-csharp_preferred_modifier_order = public,private,protected,internal,static,extern,new,virtual,abstract,sealed,override,readonly,unsafe,volatile,async:suggestion
+# ReSharper properties
```
Comment (tgiphil) [inline Source/.editorconfig:29]: Not using reshapper.

File: Source/.editorconfig [hunk 2]
```diff
@@ -11,24 +11,65 @@ indent_style = tab
 indent_size = 4
 trim_trailing_whitespace = true
 
-[*.cs]
```
Comment (tgiphil) [inline Source/.editorconfig:14]: Fix.

File: Source/Mosa.BareMetal.CoolWorld/Display.cs [hunk 3]
```diff
@@ -2,6 +2,9 @@
 
 using System.Drawing;
 using Mosa.DeviceSystem;
+using Mosa.DeviceSystem.Devices.Graphics;
+using Mosa.DeviceSystem.Drivers.Graphics;
```
Comment (tgiphil) [inline Source/Mosa.BareMetal.CoolWorld/Display.cs:6]: Duplicate

File: Source/Mosa.DeviceDriver/PCI/VMware/VMwareSVGA2.cs [hunk 4]
```diff
@@ -19,6 +19,8 @@
 
 using System;
 using Mosa.DeviceSystem;
+using Mosa.DeviceSystem.Devices.Graphics;
+using Mosa.DeviceSystem.Drivers.Graphics;
```
Comment (tgiphil) [inline Source/Mosa.DeviceDriver/PCI/VMware/VMwareSVGA2.cs:23]: Duplicate.

File: Source/Mosa.DeviceDriver/ScanCodeMap/HU.cs [hunk 5]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) MOSA Project. Licensed under the New BSD License.
 
 using Mosa.DeviceSystem;
+using Mosa.DeviceSystem.Devices.Keyboard;
+using Mosa.DeviceSystem.Drivers.Keyboard;
```
Comment (tgiphil) [inline Source/Mosa.DeviceDriver/ScanCodeMap/HU.cs:5]: Duplicate

File: Source/Mosa.DeviceDriver/ScanCodeMap/US.cs [hunk 6]
```diff
@@ -1,6 +1,8 @@
 ﻿// Copyright (c) MOSA Project. Licensed under the New BSD License.
 
 using Mosa.DeviceSystem;
+using Mosa.DeviceSystem.Devices.Keyboard;
+using Mosa.DeviceSystem.Drivers.Keyboard;
```
Comment (tgiphil) [inline Source/Mosa.DeviceDriver/ScanCodeMap/US.cs:5]: Duplicate

File: Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs [hunk 7]
```diff
@@ -39,43 +32,20 @@ public FrameBuffer32(ConstrainedPointer buffer, uint width, uint height, Func<ui
...
-	/// <summary>Sets the pixel.</summary>
-	/// <param name="color"></param>
-	/// <param name="x">The x.</param>
-	/// <param name="y">The y.</param>
 	public void SetPixel(uint color, uint x, uint y)
 	{
-		if (x >= Width || y >= Height)
-			return;
-
+		if (x >= Width || y >= Height) return;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs:39]: Separate line.

File: Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs [hunk 8]
```diff
@@ -39,43 +32,20 @@ public FrameBuffer32(ConstrainedPointer buffer, uint width, uint height, Func<ui
...
-	/// <summary>Gets the pixel.</summary>
-	/// <param name="x">The x.</param>
-	/// <param name="y">The y.</param>
-	/// <returns></returns>
 	public uint GetPixel(uint x, uint y)
 	{
-		if (x >= Width || y >= Height)
-			return 0;
-
+		if (x >= Width || y >= Height) return 0;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs:45]: Separate line.

File: Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs [hunk 9]
```diff
@@ -238,12 +178,10 @@ public void FillCircle(uint color, uint x, uint y, uint r)
 			radiusError += yChange;
 			yChange += 2;
 
-			if ((radiusError << 1) + xChange > 0)
-			{
-				x1--;
-				radiusError += xChange;
-				xChange += 2;
-			}
+			if ((radiusError << 1) + xChange <= 0) continue;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs:181]: Separate line.

File: Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs [hunk 10]
```diff
@@ -270,14 +208,12 @@ public void DrawLine(uint color, uint x1, uint y1, uint x2, uint y2)
...
-		for (var i = 0u; i < dx; i++)
-			SetPixel(x + i, y, color);
+		for (var i = 0u; i < dx; i++) SetPixel(x + i, y, color);
 	}
 
 	private void DrawVerticalLine(uint color, uint dy, uint x, uint y)
 	{
-		for (var i = 0u; i < dy; i++)
-			SetPixel(color, x, y + i);
+		for (var i = 0u; i < dy; i++) SetPixel(color, x, y + i);
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs:216]: Separate line.

File: Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs [hunk 11]
```diff
@@ -270,14 +208,12 @@ public void DrawLine(uint color, uint x1, uint y1, uint x2, uint y2)
 
 	private void DrawHorizontalLine(uint color, uint dx, uint x, uint y)
 	{
-		for (var i = 0u; i < dx; i++)
-			SetPixel(x + i, y, color);
+		for (var i = 0u; i < dx; i++) SetPixel(x + i, y, color);
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Devices/Graphics/FrameBuffer32.cs:211]: Separate line.

File: Source/Mosa.DeviceSystem/Drivers/Keyboard/IKeyboardDevice.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace Mosa.DeviceSystem.Drivers.Keyboard;
+
+/// <summary>
+/// An interface used for interacting directly with a keyboard device. It provides access to the device's raw scan codes, in a
+/// blocking or non-blocking way. Note that the device must obviously support retrieving raw scan codes (e.g. a PS/2 keyboard)
+/// </summary>
+public interface IKeyboardDevice
+{
+	byte GetScanCode(bool blocking = false);
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Drivers/Keyboard/IKeyboardDevice.cs:11]: Okay... but need to revisit this to be truly multithreaded.

File: Source/Mosa.DeviceSystem/PCI/AddressType.cs [hunk 13]
```diff
@@ -7,18 +7,7 @@ namespace Mosa.DeviceSystem.PCI;
 /// </summary>
 public enum AddressType
 {
-	/// <summary>
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/PCI/AddressType.cs:10]: Let's not remove these comments.

File: Source/Mosa.DeviceSystem/PCI/PCIDevice.cs [hunk 14]
```diff
@@ -75,164 +77,90 @@ internal struct PCICommand
...
-	/// <value>The base addresses.</value>
 	public BaseAddress[] BaseAddresses { get; private set; }
 
 	#endregion Properties
 
 	public override void Initialize()
 	{
-		pciController = base.Device.Parent.DeviceDriver as IPCIController;
+		pciController = Device.Parent.DeviceDriver as IPCIController;
+		if (pciController == null) return;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/PCI/PCIDevice.cs:129]: Separate line.

File: Source/Mosa.DeviceSystem/PCI/PCIDevice.cs [hunk 15]
```diff
@@ -75,164 +77,90 @@ internal struct PCICommand
...
 
-		var configuration = base.Device.Configuration as PCIDeviceConfiguration;
+		var configuration = Device.Configuration as PCIDeviceConfiguration;
+		if (configuration == null) return;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/PCI/PCIDevice.cs:132]: Separate line.

File: Source/Mosa.DeviceSystem/PCI/PCIDevice.cs [hunk 16]
```diff
@@ -75,164 +77,90 @@ internal struct PCICommand
...
 		{
-			byte barr = (byte)(PCIConfigurationHeader.BaseAddressRegisterBase + i * 4);
-
-			uint address = pciController.ReadConfig32(this, barr);
+			var bar = (byte)(PCIConfigurationHeader.BaseAddressRegisterBase + i * 4);
 
-			if (address == 0)
-				continue;
+			var address = pciController.ReadConfig32(this, bar);
+			if (address == 0) continue;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/PCI/PCIDevice.cs:146]: Separate line.

File: Source/Mosa.DeviceSystem/Services/DiskDeviceService.cs [hunk 17]
```diff
@@ -13,26 +16,17 @@ public override void PostEvent(ServiceEvent serviceEvent)
...
-
-			if (controller.GetTotalSectors(drive) == 0)
-				continue;
-
-			// don't mount twice
-			if (deviceService.CheckExists(device, drive))
-				return;
+			if (!controller.Open(drive)) continue;
+			if (controller.GetTotalSectors(drive) == 0) continue;
+			if (deviceService.CheckExists(device, drive)) return; // Don't mount twice
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Services/DiskDeviceService.cs:29]: Separate lines.

File: Source/Mosa.DeviceSystem/Services/PCService.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+	protected override void Initialize()
+	{
+		deviceService = ServiceManager.GetFirstService<DeviceService>();
+	}
+
+	public bool Reset()
+	{
+		var device = deviceService.GetFirstDevice<IACPI>(DeviceStatus.Online);
+		if (device == null) return false;
```
Comment (tgiphil) [inline Source/Mosa.DeviceSystem/Services/PCService.cs:23]: Separate line. (won't comment on all instances).

File: Source/.editorconfig [hunk 19]
```diff
@@ -11,24 +11,65 @@ indent_style = tab
...
-indent_size = 4
+# Microsoft .NET properties
+csharp_new_line_before_open_brace = all
+csharp_preferred_modifier_order = private, async, protected, public, internal, volatile, abstract, new, override, sealed, static, virtual, extern, readonly, unsafe, file, required:suggestion
```
Comment (tgiphil) [inline Source/.editorconfig:16]: This should not be universal.

**Review by tgiphil — APPROVED** [review]

File: Source/.editorconfig (hunk 1 above)
Comment (charsleysa) [reply Source/.editorconfig:29]: I tihnk it's fine to leave this config here, it allows code contributions from resharper to align with VS

File: Source/.editorconfig [hunk 20]
```diff
@@ -11,24 +11,65 @@ indent_style = tab
...
+resharper_place_field_attribute_on_same_line = false
+resharper_place_simple_embedded_statement_on_same_line = false
+resharper_place_simple_initializer_on_single_line = false
+resharper_space_after_ptr_in_data_member = false
+resharper_space_before_ptr_in_data_member = true
+resharper_toplevel_function_declaration_return_type_style = on_single_line
 
 file_header_template = Copyright (c) MOSA Project. Licensed under the New BSD License.
-
-[*.csproj]
```
Comment (charsleysa) [inline Source/.editorconfig:29]: why were these removed?

**Review by charsleysa — CHANGES_REQUESTED** [review]

File: Source/Mosa.BareMetal.CoolWorld/Display.cs (hunk 3 above)
Comment (charsleysa) [reply Source/Mosa.BareMetal.CoolWorld/Display.cs:6]: one is Devices.Graphics while the other is Drivers.Graphics

File: Source/Mosa.DeviceDriver/PCI/VMware/VMwareSVGA2.cs (hunk 4 above)
Comment (charsleysa) [reply Source/Mosa.DeviceDriver/PCI/VMware/VMwareSVGA2.cs:23]: one is Devices.Graphics while the other is Drivers.Graphics

File: Source/Mosa.DeviceDriver/ScanCodeMap/HU.cs (hunk 5 above)
Comment (charsleysa) [reply Source/Mosa.DeviceDriver/ScanCodeMap/HU.cs:5]: one is Devices.Keyboard while the other is Drivers.Keyboard

File: Source/Mosa.DeviceDriver/ScanCodeMap/US.cs (hunk 6 above)
Comment (charsleysa) [reply Source/Mosa.DeviceDriver/ScanCodeMap/US.cs:5]: one is Devices.Keyboard while the other is Drivers.Keyboard

File: Source/.editorconfig (hunk 20 above)
Comment (AnErrupTion) [reply Source/.editorconfig:29]: The yml one was invalid (contained an ending brace) and it wasn't used anyway, since the yml files are included starting from the parent directory (root), and not Source. Should the EditorConfig file be moved at the root instead?

Comment (charsleysa) [reply Source/.editorconfig:29]: Sounds good

**Review by tgiphil — APPROVED** [review]

**Review by charsleysa — APPROVED** [review]
LGTM
